cpp/ql/src/experimental/Security/CWE/CWE-266/IncorrectPrivilegeAssignment.ql

```diff
@@ -0,0 +1,77 @@
+/**
+ * @name Find the wrong use of the umask function.
+ * @description Incorrectly evaluated argument to the umask function may have security implications.
+ * @kind problem
+ * @id cpp/wrong-use-of-the-umask
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       maintainability
+ *       security
+ *       external/cwe/cwe-266
+ *       external/cwe/cwe-264
+ *       external/cwe/cwe-200
+ *       external/cwe/cwe-560
+ *       external/cwe/cwe-687
+ */
+
+import cpp
+import semmle.code.cpp.exprs.BitwiseOperation
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds for a function `f` that has an argument at index `apos` used to set file permissions. */
+predicate numberArgumentModFunctions(Function f, int apos) {
+  f.hasGlobalOrStdName("umask") and apos = 0
+  or
+  f.hasGlobalOrStdName("fchmod") and apos = 1
+  or
+  f.hasGlobalOrStdName("chmod") and apos = 1
+}
+
+from FunctionCall fc, string msg
+where
+  fc.getTarget().hasGlobalOrStdName("umask") and
+  fc.getArgument(0).getValue() = "0" and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("umask") and
+    globalValueNumber(fctmp.getArgument(0)) != globalValueNumber(fc.getArgument(0))
+  ) and
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("fopen") or
+      fctmp.getTarget().hasGlobalOrStdName("open")
+    ) and
+    fctmp.getNumberOfArguments() = 2 and
+    fctmp.getArgument(0).getValue() != "/dev/null"
+  ) and
+  not exists(FunctionCall fctmp |
+    fctmp.getTarget().hasGlobalOrStdName("chmod") or
+    fctmp.getTarget().hasGlobalOrStdName("fchmod")
+  ) and
+  msg = "Using umask (0) may not be safe."
+  or
+  fc.getTarget().hasGlobalOrStdName("umask") and
+  exists(FunctionCall fctmp |
+    (
+      fctmp.getTarget().hasGlobalOrStdName("chmod") or
+      fctmp.getTarget().hasGlobalOrStdName("fchmod")
+    ) and
+    (
+      globalValueNumber(fc.getArgument(0)) = globalValueNumber(fctmp.getArgument(1)) and
+      fc.getArgument(0).getValue() != "0"
+    ) and
+    msg = "not use equal argument in umask and " + fctmp.getTarget().getName() + " functions"
+  )
+  or
+  exists(Expr exptmp, int i |
+    numberArgumentModFunctions(fc.getTarget(), i) and
+    not exptmp.getAChild*() instanceof FunctionCall and
+    not exists(SizeofOperator so | exptmp.getAChild*() = so) and
+    not exists(ArrayExpr aetmp | aetmp.getArrayOffset() = exptmp.getAChild*()) and
+    exptmp.getAChild*() instanceof BinaryArithmeticOperation and
+    not exptmp.getAChild*() instanceof BinaryBitwiseOperation and
+    globalValueNumber(exptmp) = globalValueNumber(fc.getArgument(i)) and
+    not exptmp.isConstant() and
+    msg = "Using arithmetic to compute the mask may not be safe."
```

Comment:
So this case is trying to spot when permissions are calculated with arithmetic operations like `+` or `-` rather than bitwise operations like `&` and `~`.  That seems reasonable.

I'm not convinced the ad-hoc exemptions for function calls, sizeof and array operations is good design though - there are probably many other things that should be excluded if we go this route.  Perhaps just looking at the top-level `Expr` (not the children) would be simpler and catch most cases without the need for a list of exceptions?  Alternatively, we might be able to come up with a more principled recursive analysis of the `Expr`?

Reply:
thanks for that suggestion.
I always learn a lot from your comments.
I'll figure out how to do it. (if you have a link to an example of what you mean it would be great)

Reply:
Did you mean something like this when talking about recursion

```ql
class ExprRecur extends Expr {
  ExprRecur() {
    this instanceof SizeofOperator or
    exists(ArrayExpr aetmp | this = aetmp.getArrayOffset()) and
    this instanceof FunctionCall or
    this instanceof BinaryBitwiseOperation or
    // enabling recursive search
    exists(ExprRecur exprec | this = exprec.getAChild())
  }
}
```

Reply:
That's a good start, but I think we should avoid having logic for unrelated things like `SizeofOperator` and so on, since we're unlikely to produce something robust this way. Instead of recursing into all children and making exceptions, we can specify a limited class of things we *do* want to recurse into. We do seem to want to recurse into `BinaryBitwiseOperation`s looking for a `BinaryArithmeticOperation` underneath. Does the following get the results you want?
```
/**
 * An expression that is either a `BinaryArithmeticOperation` or the result of one or more `BinaryBitwiseOperation`s on a `BinaryArithmeticOperation`. For example `1 | (2 + 3)`.
 */
class ContainsArithmetic extends Expr {
  ContainsArithmetic() {
    this instanceof BinaryArithmeticOperation
    or
    // recursive search into `Operation`s
    this.(BinaryBitwiseOperation).getAnOperand() instanceof ContainsArithmetic
  }
}
```

Reply:
Good afternoon.
thank you this is really good.
i use this.